scipy/sparse/csgraph/_tools.pyx

```diff
@@ -219,6 +219,64 @@ def csgraph_from_dense(graph,
                                                          infinity_null))
 
 
+def csgraph_from_adjacency_list(adjacency_list, weighted=True):
+    """
+    csgraph_from_adjacency_list(adjacency_list, weighted=True)
+
+    Construct a CSR-format sparse graph from an adjacency list.
+
+    .. versionadded:: 1.7.2
+
+    Parameters
+    ----------
+    adjacency_list : dict
+        Input adjacency list.  Could be either a dict of lists (for unweighted graph) or
+        a dict of dicts (for weighted graph). Dict lenght should be n_nodes.
+    weighted : bool
+        If True (default), then constract a weighted graph.
+
+    Returns
+    -------
+    csgraph : csr_matrix
+        Compressed sparse representation of graph,
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import csgraph_from_adjacency_list
+
+    >>> graph = {
+    ...     0: {1: 1, 2: 2},
+    ...     1: {3: 1},
+    ...     2: {3: 3},
+    ...     3: {},
+    ... }
+
+    >>> csgraph_from_adjacency_list(graph)
+    <4x4 sparse matrix of type '<class 'numpy.int64'>'
+	    with 4 stored elements in Compressed Sparse Row format>
+    """
+
+    N = len(adjacency_list)
+
+    row = []
+    col = []
+    data = []
+
+    for row_idx in adjacency_list:
+        for col_idx in adjacency_list[row_idx]:
+            row.append(row_idx)
+            col.append(col_idx)
+
+            if weighted:
+                if not isinstance(adjacency_list[row_idx], dict):
+                    raise ValueError("if 'weighted=True', adjacency_list should be a dict of dicts.")
```

Comment:
This need to be tested with `with pytest.raises(ValueError, match="adjacency_list should"):`

Reply:
Done!